design/nfs/nfs-controller-runtime.md

```diff
@@ -0,0 +1,161 @@
+
+The current implementation is the operator watch an event (create, update and delete) of CustomResource and will be handled by registered function in `ResourceEventHandlerFuncs` which every event has its own handler but only the create handler that implemented.
+
+Controller-runtime introduces an interface called `Reconciler` that will ensure the state of the system matches what is specified by the user in the object at the time the `Reconciler` is called. `Reconciler` responds to generic events so it will contain all of the business logic of a Controller (create, update, and delete). The controller-runtime also have utility functions for creating and updating an object called `CreateOrUpdate` which will make easier to handling update of an object.
+
+Since this implementation only changes the logic of the controller, so the deployment process will be like current implementation.
+
+### Validation
+
+CustomResource validation in the operator can be done through the Controller itself. However, the operator pattern has two common types to validate the CustomResource.
+
+- **Syntactic validation** By defining OpenAPI validation rules.
+- **Semantic Validation** By creating a ValidatingAdmissionConfiguration and Admission Webhook.
+
+The current implementation only validates the CustomResource in the controller and just gives an error log in the operator stdout if the given resource is invalid. In this implementation will also cover the CustomResouce validation both though `Syntactic validation` and `Semantic Validation` and also give an improvement validation in the controller.
+
+To implement `Syntactic validation` is only by defining OpenAPI validation rules. Otherwise, the `Semantic Validation` implementation is a bit more complicated. Fortunately, controller-runtime provides an awesome package to help to implement Admission Webhook. Since the webhook server must be served through TLS, a valid TLS certificate will be required. In this case, we can depend on [cert-manager](https://cert-manager.io/)
+
+Example self signed certificate.
+
+```yaml
+---
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: rook-nfs-webhook-cert
+  namespace: rook-nfs-system
+spec:
+  dnsNames:
+  - rook-nfs-webhook.rook-nfs-system.svc
+  - rook-nfs-webhook.rook-nfs-system.svc.cluster.local
+  issuerRef:
+    kind: Issuer
+    name: rook-nfs-selfsigned-issuer
+  secretName: rook-nfs-webhook-cert
+---
+apiVersion: cert-manager.io/v1alpha2
+kind: Issuer
+metadata:
+  name: rook-nfs-selfsigned-issuer
+  namespace: rook-nfs-system
+spec:
+  selfSigned: {}
+```
+
+And the ValidatingAdmissionConfiguration will look like
+
+```yaml
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  annotations:
+    cert-manager.io/inject-ca-from: rook-nfs-system/rook-nfs-webhook-cert
+  creationTimestamp: null
+  name: rook-nfs-validating-webhook-configuration
+webhooks:
+- clientConfig:
+    caBundle: Cg==
```

Comment:
is this value `Cg==` special? i think you're saying it gets replaced because of `inject-ca-from` annotation, so does its value matter?